`[ENG-936]` introduce AccessControl to ERC20Lockable with transfer and minter roles

twodam opened the pull request:
Depends on https://github.com/decentdao/decent-contracts/pull/296

### Summary

* Replaced the previous "Transfer Whitelist" mechanism with [AccessControlUpgradeable](https://github.com/OpenZeppelin/openzeppelin-contracts-upgradeable/blob/master/contracts/access/AccessControlUpgradeable.sol).
* Simplified `isTransferable` modifier by using `hasRole` and grant transfer role to `address(0)`.
* Introduced a `MINTER_ROLE` to manage the minter whitelist.
* The `owner` is granted the [DEFAULT_ADMIN_ROLE](https://github.com/OpenZeppelin/openzeppelin-contracts-upgradeable/blob/3eee293579c9d248235acadbdc246a46b1cadeab/contracts/access/AccessControlUpgradeable.sol#L46), acting as the default admin for all roles, including transfer and minter roles.

twodam committed:
refactor: use AccessControl to replace whitelist

twodam committed:
feat(VotesERC20LockableV1): add minter role

twodam committed:
refactor(VotesERC20LockableV1): grant transfer role to zero address for enabling mint even token is locked

tbwebb22 commented on contracts/deployables/erc20/VotesERC20V1.sol:
```diff
@@ -100,12 +104,19 @@ contract VotesERC20V1 is
             interfaceId == type(IERC20).interfaceId ||
             interfaceId == type(IERC20Permit).interfaceId ||
             interfaceId == type(IVotes).interfaceId ||
+            interfaceId == type(IVersion).interfaceId ||
```
since this contract inherits from Version, we shouldn't need this line right? I think super.supportsInterface(interfaceId) should handle that for us

twodam replied:
Both parent contracts implement the `supportsInterface` method, which requires using `override(Version, AccessControlUpgradeable)`. Since `AccessControlUpgradeable` is listed last in the inheritance list, the call to `super.supportsInterface` actually resolves to `AccessControlUpgradeable.supportsInterface`.
I have updated this line to be more explicit: `Version.supportsInterface(interfaceId) ||`

tbwebb22 commented on contracts/deployables/erc20/VotesERC20V1.sol:
```diff
@@ -14,14 +15,16 @@ import {ERC20PermitUpgradeable} from "@openzeppelin/contracts-upgradeable/token/
     UUPSUpgradeable,
-    Ownable2StepUpgradeable
+    Ownable2StepUpgradeable,
+    AccessControlUpgradeable
```
Do we need AccessControl inherited in this contract? Doesn't look like we're using it here, and could just inherit it in VotesERC20LockableV1.sol

twodam replied:
Inheriting AccessControl directly in VotesERC20LockableV1 results in a [Linearization of inheritance graph impossible](https://docs.soliditylang.org/en/v0.8.30/contracts.html#multiple-inheritance-and-linearization) error. However, inheriting it in VotesERC20V1 compiles without issues. I’m not sure I fully grasp the [diamond problem](https://en.wikipedia.org/wiki/Multiple_inheritance#The_diamond_problem), but this setup works as expected.
<img width="202" alt="image" src="https://github.com/user-attachments/assets/d5e01555-2721-4ed1-8181-66b6da5724f3" />

twodam replied:
Maybe we could update `Version` to remove ERC165, similar to what @adamgall did in [this comment](https://github.com/decentdao/decent-contracts/pull/314#issuecomment-2937679873)?
though we may need to all places referencing `Version.supportsInterface`.

adamgall replied:
@twodam i agree with your suggestion. i've just created a PR which removes ERC165 inheritance from Version: https://github.com/decentdao/decent-contracts/pull/319.

After that PR is merged, merge develop into your branch here then move the AccessControl inheritance out of VotesERC20 and into VotesERC20Lockable.

twodam replied:
updated!

tbwebb22 commented on contracts/deployables/erc20/VotesERC20LockableV1.sol:
```diff
@@ -22,9 +24,7 @@ contract VotesERC20LockableV1 is IVotesERC20LockableV1, VotesERC20V1 {
         if (
             _locked &&
             // overrides while locked
-            !(from == owner() || // owner can always transfer
-                _whitelisted[from] || // whitelisted addresses can always transfer
-                from == address(0)) && // can always mint when locked
+            !hasRole(TRANSFER_ROLE, from) && // whitelisted addresses can always transfer
             to != address(0) // can always burn when locked
         ) {
             revert IsLocked();
```
I'm a little confused on how this is working with the minter role. 

Since _mint calls _update, which we've overriden with this modifier. And in this modifier we aren't checking for the MINTER_ROLE. Maybe I'm missing something, but do you mind explaining that?

twodam replied:
The MINTER_ROLE is enforced in the mint function, as shown below:
```solidity
function mint(
    address to,
    uint256 amount
) external virtual override onlyRole(MINTER_ROLE) {
```
To elaborate further: both transfer and mint ultimately call _update, but transfer explicitly forbids either from or to from being address(0). Therefore, the only valid way to mint tokens is through the mint function, which is why the MINTER_ROLE check is implemented there.

twodam committed:
feat(VotesERC20V1): make it explicit that Version.supportsInterface to be keep

twodam committed:
Merge branch 'develop' into feature/eng-936-allow-mint-through-trusted-separate-mint-whitelist-which-the

adamgall commented on contracts/deployables/erc20/VotesERC20LockableV1.sol:
```diff
@@ -5,15 +5,24 @@ import {IVotesERC20LockableV1} from "../../interfaces/decent/deployables/IVotesE
-
-contract VotesERC20LockableV1 is IVotesERC20LockableV1, VotesERC20V1 {
+import {IAccessControl} from "@openzeppelin/contracts/access/IAccessControl.sol";
+import {ERC165Upgradeable} from "@openzeppelin/contracts-upgradeable/utils/introspection/ERC165Upgradeable.sol";
```
What's the reason for using ERC165Upgradeable instead of ERC165?

twodam replied:
updated

twodam committed:
fix(VotesERC20LockableV1): use ERC165 instead of ERC165Upgradeable

twodam committed:
Merge branch 'develop' into feature/eng-936-allow-mint-through-trusted-separate-mint-whitelist-which-the

twodam committed:
fix(VotesERC20LockableV1): remove erc165_init

adamgall commented on contracts/deployables/erc20/VotesERC20LockableV1.sol:
```diff
@@ -63,14 +72,6 @@ contract VotesERC20LockableV1 is IVotesERC20LockableV1, VotesERC20V1 {
-
     function setMaxTotalSupply(
         uint256 newMaxTotalSupply
     ) external virtual override onlyOwner {
```
Do you think it's weird that we're still using `Ownable` modifiers (from the base VotesERC20 contract) in this contract? I think that's a little weird. I think I'd like to see another role defined here for "modifying state" (of which the `_owner` is added to in initializer), then use that role in place of the two `onlyOwner` modifiers in this contract (`lock()` and `setMaxTotalSupply()`).

@tbwebb22 thoughts?

tbwebb22 replied:
hmmm so the owner will pretty much always be the DAO. The idea with roles for minting and transferring is that there could be a need to have multiple addresses that can mint or transfer.

I'm struggling to see a use case for having an address other than the DAO be able to `lock()` or `setMaxSupply()`, can you think of any?

I can also understand the argument for using roles for `lock()` and `setMaxSupply()`  just for consistency however.

adamgall replied:
My reasoning is more so for consistency and separation of concerns.

The base `VotesERC20` is `Ownable` ONLY for the `_authorizeUpgrade()` call. There are no other `onlyOwner` functions.

But then in this contract, we introduce the concept of more refined Access Control, but then _also_ use `onlyOwner` for only _some_ of those authorized functions. That just feels weird to me.

I certainly don't disagree that in practice there will very likely not be a use case where someone other than the DAO is in charge of locking or changing max supply, but I think it's still strange to use `onlyOwner` on those.

Why not do `onlyRole(DEFAULT_ADMIN_ROLE)` for those two functions instead of `onlyOwner`? I think that would achieve the same result, and it would also fully remove `onlyOwner` from this contract.

tbwebb22 replied:
ok yeah makes sense to me to maintain consistency, so I agree we should add these two roles

adamgall replied:
I've implemented this

adamgall committed:
Merge branch 'develop' into feature/eng-936-allow-mint-through-trusted-separate-mint-whitelist-which-the

adamgall committed:
refactor(VotesERC20LockableV1): update parameter naming and function signatures

**Summary of Changes:**

- Updated function parameter names in `VotesERC20LockableV1` and `IVotesERC20LockableV1` to follow the convention of appending an underscore (`_`) to parameter names for clarity.
- Modified the `maxTotalSupply`, `setMaxTotalSupply`, `mint`, `burn`, and `supportsInterface` functions to reflect the new parameter naming convention.
- Ensured that all functions in the contract are marked as `virtual` and `override` where applicable, enhancing compatibility with interface definitions.

These changes improve code readability and maintainability by adhering to established naming conventions and ensuring consistency across the contract and its interface.

adamgall committed:
refactor(VotesERC20LockableV1): update access control for lock and setMaxTotalSupply functions

**Summary of Changes:**

- Updated the `lock` and `setMaxTotalSupply` functions in `VotesERC20LockableV1` to use `onlyRole(DEFAULT_ADMIN_ROLE)` instead of `onlyOwner`, enhancing access control by allowing only users with the default admin role to execute these functions.
- Adjusted the corresponding test cases to reflect the change in error messages from `OwnableUnauthorizedAccount` to `AccessControlUnauthorizedAccount`, ensuring that the tests accurately validate the new access control logic.

These changes improve the security and flexibility of the contract by aligning access control with role-based permissions.
